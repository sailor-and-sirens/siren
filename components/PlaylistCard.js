import React, { Component } from 'react';
import { StyleSheet, Text, View, Image, TouchableOpacity} from 'react-native';
import { connect } from 'react-redux';
import Swipeable from 'react-native-swipeable';
import { actionCreators } from '../actions';
import { headerActions } from '../actions/Header';
import { actionCreators as swipeActions } from '../actions/Swipe';
import { actionCreators as playlistActions } from '../actions/Playlist';

const mapStateToProps = (state) => ({
  token: state.main.token,
  rightActionActivated: state.swipe.isRightActionActivated,
});

class PlaylistCard extends Component {

  componentWillMount(){
    this.imagesArr = [];
    this.setPlaylistImages();
  }

  removePlaylist(playlistId){
    fetch("http://siren-server.herokuapp.com/api/playlists/remove-playlist", {
      method: "DELETE",
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Authorization': this.props.token
      },
      body: JSON.stringify({playlistId: playlistId})
    })
    .catch(console.log);
  }

  convertMinutesToHrsMinutes(minutes) {
    let hours = Math.floor(minutes / 60);
    minutes = minutes - (hours * 60);
    return hours + 'h' + ' ' + minutes + 'm';
  }

  setPlaylistImages(){
    var placeholderImage = 'https://render.fineartamerica.com/images/rendered/small/print/images-square-real-5/soothing-sea-abstract-painting-linda-woods.jpg';
    this.props.playlist.Episodes.forEach(episode => {
      if (this.imagesArr.every(url => url !== episode.Podcast.artworkUrl) && this.imagesArr.length < 4){
        this.imagesArr.push(episode.Podcast.artworkUrl);
      }
    });
    if (this.imagesArr.length === 4) {
      this.imageClass = 'quad';
    } else if (this.imagesArr.length === 3) {
      this.imageClass = 'quad';
      this.imagesArr.push(this.imagesArr[0]);
    } else if (this.imagesArr.length === 2) {
      this.imageClass = 'quad';
      this.imagesArr.splice(1, 0, this.imagesArr[1]);
      this.imagesArr.splice(3, 0, this.imagesArr[0]);
    } else if (this.imagesArr.length < 1) {
      this.imageClass = 'single';
      this.imagesArr.push(placeholderImage);
    } else {
      this.imageClass = 'single';
      this.imagesArr = this.imagesArr.slice(0,1);
    }
  }

  render() {
    const {rightActionActivated} = this.props;
    return (
      <Swipeable
        rightActionActivationDistance={200}
        rightContent={(
          <View style={[styles.rightSwipeItem, {backgroundColor: rightActionActivated ? '#114B5F' : '#D62828'}]}>
            {rightActionActivated ?
              <Text style={styles.swipeText}>(( release ))</Text> :
              <Text style={styles.swipeText}>Remove Playlist</Text>}
          </View>
        )}
        onRightActionActivate={() => this.props.dispatch(swipeActions.updateRightActivation(true))}
        onRightActionDeactivate={() => this.props.dispatch(swipeActions.updateRightActivation(false))}
        onRightActionComplete={() => {
          this.removePlaylist(this.props.playlist.id);
          this.props.dispatch(playlistActions.removePlaylist(this.props.playlist.id));

        }}
      >
      <View style={styles.cardContainer}>
        <View style={styles.image}>
          <TouchableOpacity onPress={() => {
            this.props.dispatch(actionCreators.updatePlaylistFilter(this.props.playlist.name));
            this.props.dispatch(headerActions.changeView(this.props.playlist.name + ' Playlist'));
          }}>
            <View style={styles.image}>
              {this.imagesArr.map((image, index) => <Image key={index} source={{uri: image}} style={styles[this.imageClass]}/>)}
            </View>
          </TouchableOpacity>
        </View>
        <View style={styles.content}>
          <Text style={styles.playlistName}>{this.props.playlist.name}</Text>
          {this.props.playlist.Episodes.length > 0 ?
            <Text style={styles.recentlyAdded}>Recently Added:</Text> : <View></View>
          }
          {this.props.playlist.Episodes.slice(0,2).map((episode, index) => {
            return (
              <Text key={index} style={styles.episodeTitle} numberOfLines={1}>{episode.title}</Text>
            )
          })}
        </View>
        <View style={styles.timeContainer}>
          <Text style={styles.time}>{this.convertMinutesToHrsMinutes(this.props.playlist.totalTime)}</Text>
        </View>
      </View>
    </Swipeable>
    );
  }

}

const styles = StyleSheet.create({
  cardContainer: {
    height: 90,
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingTop: 5,
    borderBottomWidth: 1,
    borderColor: 'lightgrey'
  },
  image: {
    height: 80,
    width: 80,
    flexDirection: "row",
    flexWrap: "wrap",
    marginBottom: 10,
  },
  quad: {
    height: 40,
    width: 40,
  },
  single: {
    height: 80,
    width: 80,
  },
  content: {
    flex: .60,
    height: 80,
    marginLeft: 5,
    paddingRight: 5,
    justifyContent: 'center'
  },
  playlistName: {
    fontWeight: "500",
    fontSize: 16,
    marginBottom: 5
  },
  recentlyAdded: {
    fontWeight: "500",
    fontSize: 12
  },
  episodeTitle: {
    fontSize: 12
  },
  timeContainer: {
    height: 80,
    justifyContent: 'center',
  },
  time: {
    fontWeight: "400",
    fontSize: 14,
    marginRight: 5,
    color: 'grey',
  },
  rightSwipeItem: {
    flex: 1,
    justifyContent: 'center',
    paddingLeft: 20
  },
  swipeText: {
    color: '#ffffff'
  }
});

export default connect(mapStateToProps)(PlaylistCard);
