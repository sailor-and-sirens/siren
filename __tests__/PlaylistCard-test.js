import React from 'react';
import PlaylistCard from '../components/PlaylistCard.js';
import * as renderer from "react-test-renderer";
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from '../reducers/index';
const store = createStore(reducer);

var playlist = {
  title: 'Test playlist',
  Episodes: [{title: 'one episode', Podcast: {artworkUrl: 'test.com'}}],
  totalTime: 100
}

test('renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <PlaylistCard playlist={playlist} />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
