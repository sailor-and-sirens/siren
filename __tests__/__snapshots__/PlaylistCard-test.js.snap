// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  leftActionActivationDistance={125}
  leftActionReleaseAnimationConfig={null}
  leftActionReleaseAnimationFn={null}
  leftButtonWidth={75}
  leftButtonsActivationDistance={75}
  leftButtonsCloseReleaseAnimationConfig={null}
  leftButtonsCloseReleaseAnimationFn={null}
  leftButtonsOpenReleaseAnimationConfig={null}
  leftButtonsOpenReleaseAnimationFn={null}
  onLayout={[Function]}
  onLeftActionActivate={[Function]}
  onLeftActionComplete={[Function]}
  onLeftActionDeactivate={[Function]}
  onLeftActionRelease={[Function]}
  onLeftButtonsActivate={[Function]}
  onLeftButtonsCloseComplete={[Function]}
  onLeftButtonsCloseRelease={[Function]}
  onLeftButtonsDeactivate={[Function]}
  onLeftButtonsOpenComplete={[Function]}
  onLeftButtonsOpenRelease={[Function]}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onPanAnimatedValueRef={[Function]}
  onRef={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onRightActionActivate={[Function]}
  onRightActionComplete={[Function]}
  onRightActionDeactivate={[Function]}
  onRightActionRelease={[Function]}
  onRightButtonsActivate={[Function]}
  onRightButtonsCloseComplete={[Function]}
  onRightButtonsCloseRelease={[Function]}
  onRightButtonsDeactivate={[Function]}
  onRightButtonsOpenComplete={[Function]}
  onRightButtonsOpenRelease={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  onSwipeComplete={[Function]}
  onSwipeMove={[Function]}
  onSwipeRelease={[Function]}
  onSwipeStart={[Function]}
  rightActionActivationDistance={200}
  rightActionReleaseAnimationConfig={null}
  rightActionReleaseAnimationFn={null}
  rightButtonWidth={75}
  rightButtonsActivationDistance={75}
  rightButtonsCloseReleaseAnimationConfig={null}
  rightButtonsCloseReleaseAnimationFn={null}
  rightButtonsOpenReleaseAnimationConfig={null}
  rightButtonsOpenReleaseAnimationFn={null}
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
    ]
  }
  swipeReleaseAnimationConfig={
    Object {
      "duration": 250,
      "easing": [Function],
      "toValue": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
  swipeReleaseAnimationFn={[Function]}
  swipeStartMinDistance={15}
>
  <View
    style={
      Object {
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 0.5,
          },
        ],
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "height": 80,
              "marginBottom": 10,
              "width": 80,
            },
          ]
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "flexWrap": "wrap",
                "height": 80,
                "marginBottom": 10,
                "width": 80,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "test.com",
                }
              }
              style={
                Object {
                  "height": 80,
                  "width": 80,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 0.6,
              "marginLeft": 5,
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontWeight": "500",
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          one episode...
        </Text>
      </View>
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "grey",
                "fontSize": 14,
                "fontWeight": "400",
                "marginRight": 5,
              },
            ]
          }
        >
          1h 40m
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "marginRight": -0,
        "transform": Array [
          Object {
            "translateX": 0.5,
          },
        ],
        "width": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "center",
            "paddingLeft": 20,
          },
          Object {
            "backgroundColor": "rgb(221, 95, 95)",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Remove Playlist
      </Text>
    </View>
  </View>
</View>
`;
